lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_converter.cpp

```diff
@@ -714,6 +714,14 @@ void MmeNasStateConverter::proto_to_ue_mm_context(
   // Initialize timers to INVALID IDs
   state_ue_mm_context->mobile_reachability_timer.id = MME_APP_TIMER_INACTIVE_ID;
   state_ue_mm_context->implicit_detach_timer.id     = MME_APP_TIMER_INACTIVE_ID;
+  state_ue_mm_context->mobile_reachability_timer.msec =
```

Comment:
what is the reason to update these on state recovery? per my understanding these are set as inactive timers as we are not saving timers in redis. tagging @ulaskozat to keep me honest here.

Reply:
For registered users, when they switch to idle mode, we should be saving the timestamp and upon restart we should update it with the residual time.

Reply:
so yes, we do not need to save or recover this here. Lines 734-741 are the relevant states to recover.

Reply:
looks benign though.

Reply:
Hi Alex.. mobile_reachability_timer.msec contains the maximum time for the mobile reachability timer to expire and calculated completely on hardcoded configuration values from files mme.conf.template and mme_app_ue_context.h. This is used as a pre-configured value when starting the mobile reachability timer and has no dependency on whether the timer is inactive or active at that point. If you see the below code from file "tasks/mme_app/mme_app_location.c", this value of mobile_reachability_timer.msec is being set but the mobile_reachability_timer is still inactive. 

```
  ue_mm_context->mobile_reachability_timer.id = MME_APP_TIMER_INACTIVE_ID;
  ue_mm_context->implicit_detach_timer.id     = MME_APP_TIMER_INACTIVE_ID;
#if !MME_UNIT_TEST
  ue_mm_context->mobile_reachability_timer.msec =
      ((mme_config.nas_config.t3412_min) +
       MME_APP_DELTA_T3412_REACHABILITY_TIMER) *
      60000;
  ue_mm_context->implicit_detach_timer.msec =
      (ue_mm_context->mobile_reachability_timer.msec) +
      MME_APP_DELTA_REACHABILITY_IMPLICIT_DETACH_TIMER * 60000;

```

Now for the point you said "these are set as inactive timers as we are not saving timers in redis", the timer id not stored in redis db and set to inactive when restored after mme restart is because the timer id is generated only after we start a timer. So, when MME comes up we need to restart the timer if it was running earlier and at that time the timer ID will be freshly regenerated. But this is true for timer IDs only. For timer msec value, if you see for most of the timers it is already set here in this file starting from line no. 725 to 732

Just for additional information, the timer id with active or inactive flag is not an identifier for understanding whether a timer was running or not before mme restart. The actual identifier here are timer_started variables (e.g., state_ue_mm_context->time_mobile_reachability_timer_started), which is zero until a timer is strted and timer start time is capture in this variable.

Reply:
> For registered users, when they switch to idle mode, we should be saving the timestamp and upon restart we should update it with the residual time.

Hi Ulas.. We don't update any of these variables in the MME code with the "residual time", The (timer).msec variale will always hold the maximum timer value for the timer to expire and (timer)_started variable will always hold the time when the timer was actually started and same value will be restored till the timer expires even after MME restart. What actually changes is that we create a temporary variable to store the remaining time and call the StartTimer function to start the timer with this newly created variable. Please see the code segment below to resume timer from file "tasks/mme_app/mme_app_timer_management.cpp"

```
  time_t current_time       = time(NULL);
  time_t elapsed_time_in_ms = (current_time - start_time) * 1000;
  uint32_t remaining_time_in_msecs = timer->msec - elapsed_time_in_ms;

  magma::lte::MmeUeContext::Instance().StartTimer(
           remaining_time_in_msecs, TIMER_REPEAT_ONCE, timer_expiry_handler,
           arg)) == -1) 
```

> so yes, we do not need to save or recover this here. Lines 734-741 are the relevant states to recover.

Since we use this (timer).msec value to calculate the remaining time for the timer to run, we need to recover this (timer).msec value somewhere before actually resuming the timer.. Either in this function or in the function to resume timer after verifying the timer is started or not.

> looks benign though as it will be reset during resume_timer.

It is not being reset anywhere as it was set at only one place ,i.e., after ULA in the code "tasks/mme_app/mme_app_location.c" so far. I have added it here and not in resume timer function to have this value stored even if MME was restarted when the UE was in active mode. Otherwise this would only serve if MME was restarted when UE was in IDLE mode.

Reply:
anyhow, in general it is a good idea to initialize these rather than leaving them dangling (cc @ardzoht)

Reply:
Yes Ulas.. This mobile reachability timer/implicit detach timer will not be running while UE is in active mode.. But (timer).msec would have been initialized for these timers before MME restart during ULA and after MME restart (timer).msec value will not be re-initialized if UE is in active mode.. So after MME restart, if UE goes to idle mode, it will try to use this value to start the timer, but it will fail since this is not set. This is why restoring this data here will serve both scenarios of MME restart whether UE is in active mode or idle mode before MME restart..